File: azw3r.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

int main(int argc, char *argv[])
{
  char *s, *inpnam=NULL, *rawnam=NULL, notestr[16384], numstr[16];
  unsigned char *buf, *dat, *data;
  int fdinp, fdraw, highlight=0, note=0, offset=14;
  unsigned hbeg, hblen, hend, helen, nbeg, nblen, nend, nelen, flen;
  struct stat sbuf;
  FILE *fhraw=NULL;

  s = *argv++;
  while ((s = *argv++)) {
    if (*s == '-') switch(*++s) {
      case 'i': inpnam = *argv++; break;
      case 'h': highlight = 1; break;
      case 'n': note = 1; break;
      case 'o': offset = atoi(*argv++); break;
      case 'r': rawnam = *argv++; break;
    }
  }
  if (rawnam) fhraw = fopen(rawnam, "r");
  if (!(highlight || note)) note = 1;
  if (stat(inpnam, &sbuf)) {
    fprintf(stderr, "Can't stat input file: '%s'!\n", inpnam);
    exit(1);
  }
  if ((fdinp = open(inpnam, O_RDONLY)) < 0) {
    fprintf(stderr, "Can't open input file: '%s'!\n", inpnam);
    exit(1);
  }
  dat = buf = mmap(0, sbuf.st_size, PROT_READ, MAP_SHARED, fdinp, 0);
  while(dat - buf < sbuf.st_size) {
    if (highlight) if (!strncmp(dat, "annotation.personal.highlight", 29)) {
      if (dat[29] == 3) {
        dat += 30;
        hblen = (dat[0]*256 + dat[1])*256 + dat[2]; dat += 3;
        strncpy(numstr, dat, hblen); numstr[hblen] = 0;
        sscanf(numstr, "%u", &hbeg); dat += hblen;
      } else {
        fprintf(stderr, "File synchronization lost.\n");
        goto NEXTbyte;
      }
      if (dat[0] == 3) {
        dat++;
        helen = (dat[0]*256 + dat[1])*256 + dat[2]; dat += 3;
        strncpy(numstr, dat, helen); numstr[helen] = 0;
        sscanf(numstr, "%u", &hend); dat += helen;
      } else {
        fprintf(stderr, "File synchronization lost.\n");
        goto NEXTbyte;
      }
      if (fhraw) { int n;
        fseek(fhraw, hbeg + offset, SEEK_SET);
        n = fread(notestr, 1, hend - hbeg + 1, fhraw);
        notestr[n] = 0;
      }
      printf("%u\t%u\tHighlight:", hbeg, hend);
      if (fhraw) { printf("\t'%s'\n", notestr); } else printf("\n");
    }
    if (note) if (!strncmp(dat, "annotation.personal.note", 24)) {
      if (dat[24] == 3) {
        dat += 25;
        nblen = (dat[0]*256 + dat[1])*256 + dat[2]; dat += 3;
        strncpy(numstr, dat, nblen); numstr[nblen] = 0;
        sscanf(numstr, "%u", &nbeg); dat += nblen;
      } else {
        fprintf(stderr, "File syncnronization lost.\n");
        goto NEXTbyte;
      }
      if (dat[0] == 3) {
        dat++;
        nelen = (dat[0]*256 + dat[1])*256 + dat[2]; dat += 3;
        strncpy(numstr, dat, nelen); numstr[nelen] = 0;
        sscanf(numstr, "%u", &nend); dat += nelen;
      } else {
        fprintf(stderr, "File syncnronization lost.\n");
        goto NEXTbyte;
      }
      if (dat[0] == 2) {
        dat++;
        flen = (dat[0]*256 + dat[1])*256 + dat[2]; dat += 3; dat += flen;
        BadB2: while (dat[0] != 2) dat++;
        dat++;
        if (dat[0]) goto BadB2;
        flen = (dat[0]*256 + dat[1])*256 + dat[2]; dat += 3; dat += flen;
        while (dat[0] != 3) dat++;
        dat++;
        flen = (dat[0]*256 + dat[1])*256 + dat[2]; dat += 3; dat += flen;
        while (dat[0] != 3) dat++;
        dat++;
        flen = (dat[0]*256 + dat[1])*256 + dat[2]; dat += 3;
        if (flen > 16383) flen = 16383;
        strncpy(notestr, dat, flen); notestr[flen] = 0;
      } else {
        fprintf(stderr, "File syncnronization lost.\n");
        goto NEXTbyte;
      }
      printf("%u\t%u\tNote:\t'%s'\n", nbeg, nend, notestr);
    }
    NEXTbyte: dat++;
  }
}

File: azw3r.pl
#!/usr/bin/perl
use strict;
use warnings;
use Getopt::Long;

my ($inp, $buf, $loc, $n, $hl, $raw); my $highlight=0; my $note=0; my $pos=0;
my $b00 = pack 'C3', (2, 0, 0); my $c00 = pack 'C3', (3, 0, 0);
my $c0 = pack 'C2', (3, 0); my $offset = 0;
GetOptions(
           "inp=s" => \$inp,
           "h"     => \$highlight,
           "n"     => \$note,
           "offset=i" => \$offset,
           "raw=s" => \$raw
          );
if ($raw) { open FHraw, $raw or die "Can't open $raw for reading"; }
$note = 1 unless $highlight or $note;
my ($dev, $ino, $mode, $nlink, $uid, $gid, $rdev, $size, @misc) = stat($inp)
or die "Can't stat input file: '$inp'.\n";
open(FHazw3r, "<", $inp) or die "Can't open input file: '$inp'.\n";
$n = read FHazw3r, $buf, $size;
while ($pos < $size) {
  if ($highlight) {
    if (($loc = index $buf, 'annotation.personal.highlight', $pos) > -1) {
      warn "File synch lost looking for highlight beginning."
        if unpack('C', substr($buf, $loc + 29, 1)) != 3;
      my $hblen = unpack('n', substr($buf, $loc + 31, 2));
      my $hbeg = substr($buf, $loc + 33, $hblen); $pos = $loc + 33 + $hblen;
      warn "File synch lost looking for highlight end."
        if unpack('C', substr($buf, $pos, 1)) != 3;
      my $helen = unpack('n', substr($buf, $pos + 2, 2));
      my $hend = substr($buf, $pos + 4, $helen); $pos = $pos + 4 + $helen;
      if ($raw) {
        seek FHraw, $hbeg + $offset, 0;
        $n = read FHraw, $hl, $hend - $hbeg + 1;
      }
      print "$hbeg\t$hend\tHighlight:";
      if ($raw) { print "\t'$hl'\n"; } else { print "\n"; }
    }
    
  }
  if ($note) {
    if (($loc = index $buf, 'annotation.personal.note', $pos) > -1) {
      warn "File synch lost looking for note beginning."
        if unpack('C', substr($buf, $loc + 24, 1)) != 3;
      my $nblen = unpack('n', substr($buf, $loc + 26, 2));
      my $nbeg = substr($buf, $loc + 28, $nblen); $pos = $loc + 28 + $nblen;
      warn "File synch lost looking for note end."
        if unpack('C', substr($buf, $pos, 1)) != 3;
      my $nelen = unpack('n', substr($buf, $pos + 2, 2));
      my $nend = substr($buf, $pos + 4, $nelen); $pos = $pos + 4 + $nelen;
      $loc = index $buf, $b00, $pos;
      my $flen = unpack('n', substr($buf, $loc+2, 2)); $pos = $loc+2 + $flen;
      $loc = index $buf, $b00, $pos;
      $flen = unpack('n', substr($buf, $loc+2, 2)); $pos = $loc+2 + $flen;
      $loc = index $buf, $c00, $pos;
      $flen = unpack('n', substr($buf, $loc+2, 2)); $pos = $loc+2 + $flen;
      $loc = index $buf, $c0, $pos;
      $flen = unpack('n', substr($buf, $loc+2, 2)); $pos = $loc+4;
      my $notestr = substr $buf, $pos, $flen;
      print "$nbeg\t$nend\tNote:\t'$notestr'\n";
    }
  }
  $pos ++;
}
